src/hazelweb/gui/ActionPanel.re

```diff
@@ -0,0 +1,501 @@
+          [text("Insert "), mono_text("[] (nil)")],
+        ),
+        combo(Semicolon, simple("Cons operator")),
+        combo_and_cursor(
+          ~on_type=true,
+          LeftBracket,
+          simple("Insert type List"),
+        ),
+      ],
+    ),
+    section(
+      "Sum Types",
+      [
+        combo(Alt_L, simple("Left injection")),
+        combo(Alt_R, simple("Right injection")),
+        combo(VBar, simple("Insert | type operator")),
+      ],
+    ),
+    section(
+      "Functions",
+      [
+        combo(Backslash, simple("Insert lambda function")),
+        combo(Space, simple("Apply function")),
+        combo_and_cursor(~on_type=true, GT, [text("Create an arrow type")]),
+      ],
+    ),
+    section("Tuples", [combo(Comma, simple("Create a tuple"))]),
+    section(
+      "Pattern Matching",
+      [
+        info([
+          text("Type \""),
+          mono_text("case "),
+          text("\" to add a case expression"),
+        ]),
+        combo(Alt_C, simple("Insert case expression")),
+        combo(Enter, simple("Add new rule")),
+      ],
+    ),
+  ];
+};
+
+let view = (~inject: Update.Action.t => Vdom.Event.t, model: Model.t) => {
+  let edit_state = Model.get_edit_state(model);
+  let cursor_info = Model.get_cursor_info(model);
+
+  let is_action_allowed = (a: Action.t): bool => {
+    switch (Action.Exp.syn_perform(Contexts.empty, a, edit_state)) {
+    | Failed => false
+    | CursorEscaped(_)
+    | Succeeded(_) => true
+    };
+  };
+
+  let body = generate_panel_body(is_action_allowed, cursor_info, inject);
+
+  action_panel(body);
+};
+
+/* This function is unused at runtime, its purpose is to catch
+  * new cases added to the Acion.t type, but forgotten about in this
+  * side pane. If you add a new action, please update the code above
+  * inside generate_panel_body with a description of the new action.
+  * Afterwards the below function can be updated to not error anymore.
+  *
+  * Also consider looking at Cell.re to see if a keyboard shortcut
+  * should be added for that particular action as well.
+ */
+[@warning "-32"]
+let check_actions = (a: Action.t) =>
+  switch (a) {
+  /* Used */
+  | Backspace
+  | Delete
+  | MoveToPrevHole
+  | MoveToNextHole
+  | Construct(SOp(SArrow))
+  | Construct(SOp(SGreaterThan))
+  | Construct(SOp(SAnd))
+  | Construct(SOp(SOr))
+  | Construct(SParenthesized)
+  | Construct(SAsc)
+  | Construct(SOp(SEquals))
+  | Construct(SLine)
+  | Construct(SLam)
+  | Construct(SOp(SPlus))
+  | Construct(SOp(SMinus))
+  | Construct(SOp(STimes))
+  | Construct(SOp(SDivide))
+  | Construct(SOp(SLessThan))
+  | Construct(SOp(SSpace))
+  | Construct(SOp(SComma))
+  | Construct(SList)
+  | Construct(SListNil)
+  | Construct(SOp(SCons))
+  | Construct(SInj(L))
+  | Construct(SInj(R))
+  | Construct(SCase)
+  | Construct(SLet)
+  | Construct(SOp(SVBar))
+  | Construct(SChar(_))
+  | SwapUp
+  | SwapDown
+  | SwapLeft
+  | SwapRight
+  | MoveLeft
+  | MoveRight => ()
```

Comment:
let's define a type with one constructor, `Added`, instead of using `()`, to avoid people just reflexively copying. also let's do one switch rule per action, instead of or-ing them together like this.